File: tests/robot-java/src/main/java/frc/robot/subsystems/Elevator.java
package frc.robot.subsystems;

import com.ctre.phoenix6.CANBus;
import com.ctre.phoenix6.StatusSignal;
import com.ctre.phoenix6.controls.DutyCycleOut;
import com.ctre.phoenix6.controls.PositionVoltage;
import com.ctre.phoenix6.hardware.TalonFX;
import com.ctre.phoenix6.sim.TalonFXSimState;
import org.wpilib.command2.SubsystemBase;
import org.wpilib.framework.RobotBase;
import org.wpilib.math.system.DCMotor;
import org.wpilib.math.util.Units;
import org.wpilib.simulation.ElevatorSim;
import org.wpilib.smartdashboard.SmartDashboard;
import org.wpilib.system.RobotController;
import org.wpilib.units.measure.Angle;

public class Elevator extends SubsystemBase {
  private static final double kP = 4;
  private static final double kI = 0.0;
  private static final double kD = 0.0;

  private static final double kElevatorGearing = 10.0;
  private static final double kElevatorDrumRadius = Units.inchesToMeters(2);
  private static final double kCarriageMass = 4.0;

  private static final DCMotor kElevatorGearbox = DCMotor.getVex775Pro(4);
  private static final double kMinElevatorHeight = Units.inchesToMeters(0);
  private static final double kMaxElevatorHeight = Units.inchesToMeters(50);

  private final TalonFX m_motor;

  // Control
  private final PositionVoltage m_positionControl;
  private final DutyCycleOut m_dutyCycleControl;

  // Signals
  private final StatusSignal<Angle> m_position;

  // Sim
  private TalonFXSimState m_motorSim;
  private ElevatorSim m_elevatorSim;

  /** Create a new elevator subsystem. */
  @SuppressWarnings("this-escape")
  public Elevator() {
    m_motor = new TalonFX(PortMap.kElevatorMotorPort, new CANBus());
    m_position = m_motor.getPosition();

    m_positionControl = new PositionVoltage(0);
    m_dutyCycleControl = new DutyCycleOut(0);

    if (RobotBase.isSimulation()) {
      m_motorSim = m_motor.getSimState();
      m_elevatorSim =
          new ElevatorSim(
              kElevatorGearbox,
              kElevatorGearing,
              kCarriageMass,
              kElevatorDrumRadius,
              kMinElevatorHeight,
              kMaxElevatorHeight,
              true,
              0);
    }
  }

  public void log() {
    SmartDashboard.putNumber("Elevator Height", m_position.getValueAsDouble());
  }

  public void setVoltage(double output) {
    m_motor.setControl(m_dutyCycleControl.withOutput(output));
  }

  public void goToHeight(double height) {
    m_positionControl.Position = height;
    m_motor.setControl(m_positionControl);
  }

  public boolean isAtHeight() {
    double error = m_positionControl.Position - m_motor.getPosition().getValueAsDouble();
    return error < 0.5;
  }

  @Override
  public void periodic() {
    log();
  }

  @Override
  public void simulationPeriodic() {
    m_elevatorSim.setInput(
        m_motor.getDutyCycle().getValueAsDouble() * RobotController.getInputVoltage());
    m_elevatorSim.update(0.02);
    m_motorSim.setRawRotorPosition(m_elevatorSim.getPosition());
  }

  public void stop() {
    m_motor.setControl(m_dutyCycleControl.withOutput(0));
  }
}

File: tests/robot-cpp/subsystems/elevator.cpp
#include "robot-cpp/subsystems/elevator.hpp"

#include <wpi/math/controller/PIDController.hpp>
#include <wpi/smartdashboard/SmartDashboard.hpp>
#include <wpi/system/RobotController.hpp>

#include "robot-cpp/subsystems/ports.hpp"

namespace {
constexpr double kP = 5.0;
constexpr double kI = 0.0;
constexpr double kD = 0.0;
constexpr wpi::units::volt_t kGravityOffset{0.85};

constexpr double kElevatorGearing = 10.0;
constexpr wpi::units::meter_t kElevatorDrumRadius = 2_in;
constexpr wpi::units::kilogram_t kCarriageMass = 4.0_kg;

constexpr wpi::units::meter_t kMinElevatorHeight = 0_in;
constexpr wpi::units::meter_t kMaxElevatorHeight = 50_in;

wpi::math::DCMotor kElevatorGearbox = wpi::math::DCMotor::Vex775Pro(4);

wpi::units::meter_t TurnsToMeters(wpi::units::turn_t rotations) {
  return rotations * kElevatorDrumRadius / 1_tr;
}
wpi::units::turn_t MetersToTurns(wpi::units::meter_t meters) {
  return meters / (kElevatorDrumRadius / 1_tr);
}
}  // namespace

Elevator::Elevator()
    : m_motor{kElevatorMotorPort, ctre::phoenix6::CANBus{}},
      m_positionControl{0_tr},
      m_position(m_motor.GetPosition()),
      m_motorSim(m_motor.GetSimState()),
      m_elevatorSim(kElevatorGearbox, kElevatorGearing, kCarriageMass,
                    kElevatorDrumRadius, kMinElevatorHeight, kMaxElevatorHeight,
                    true, wpi::units::meter_t{0}) {
  ctre::phoenix6::configs::TalonFXConfiguration configs{};
  configs.Slot0.kP = kP;
  configs.Slot0.kI = kI;
  configs.Slot0.kD = kD;
  m_motor.GetConfigurator().Apply(configs);
}

void Elevator::Log() {
  wpi::SmartDashboard::PutNumber("Elevator Height (m)",
                                 GetElevatorHeight().to<double>());
}

wpi::units::meter_t Elevator::GetElevatorHeight() {
  return TurnsToMeters(m_position.GetValue());
}

void Elevator::SetVoltage(double output) {
  m_motor.SetVoltage(kGravityOffset + wpi::units::volt_t(output));
}

void Elevator::GoToHeight(wpi::units::meter_t height) {
  m_positionControl.Position = MetersToTurns(height);
  m_motor.SetControl(m_positionControl);
}

bool Elevator::IsAtHeight() {
  wpi::units::meter_t error =
      TurnsToMeters(m_positionControl.Position) - GetElevatorHeight();
  return error < 2_in;
}

void Elevator::Periodic() { Log(); }

void Elevator::SimulationPeriodic() {
  m_elevatorSim.SetInput(Eigen::Vector<double, 1>(
      m_motor.GetMotorVoltage().GetValue().to<double>()));
  m_elevatorSim.Update(20_ms);
  m_motorSim.SetRawRotorPosition(MetersToTurns(m_elevatorSim.GetPosition()));
}

void Elevator::Stop() { m_motor.SetVoltage(0_V); }
